drivers/modem/modem_iface_uart.h

```diff
@@ -55,17 +51,35 @@ int modem_iface_uart_init_dev(struct modem_iface *iface,
 			      const struct device *dev);
 
 /**
- * @brief  Init modem interface for UART
+ * @brief Initialize modem interface for UART
  *
- * @param  *iface: modem interface to initialize.
- * @param  *data: modem interface data to use
- * @param  *dev_name: name of the UART device to use
+ * @param iface Interface structure to initialize
+ * @param data UART data structure used by the modem interface
+ * @param rx_rb_buf Buffer used for internal ring buffer
+ * @param rx_rb_buf_len Size of buffer used for internal ring buffer
+ * @param dev UART device used for interface
+ * @param hw_flow_control Set if hardware flow control is used
  *
- * @retval 0 if ok, < 0 if error.
+ * @return -EINVAL if any argument is invalid
+ * @return 0 if successful
+ */
+int modem_iface_uart_init(struct modem_iface *iface, struct modem_iface_uart_data *data,
+			      char *rx_rb_buf, size_t rx_rb_buf_len, const struct device *dev,
+			      bool hw_flow_control);
+
+/**
+ * @brief Wait for rx data ready from uart interface
+ *
+ * @param iface Interface to wait on
+ *
+ * @return 0 if data is ready
+ * @return -EBUSY if returned without waiting
+ * @return -EAGAIN if timeout occurred
  */
-int modem_iface_uart_init(struct modem_iface *iface,
-			  struct modem_iface_uart_data *data,
-			  const struct device *dev);
+static inline int modem_iface_uart_rx_wait(struct modem_iface *iface, k_timeout_t timeout)
+{
+	return k_sem_take(&((struct modem_iface_uart_data *)(iface->iface_data))->rx_sem, timeout);
```

Comment:
The cast that happens here is ugly but I understand why it has to occur. While the UART interface is the only one that exists today, it seems to me that this sort of rx semaphore belongs in `struct modem_iface` rather than as an implementation detail for UART specifically. Not really sure if this is worth changing at the moment, but might be an abstraction worth adding to the modem context header for all potential interfaces going forwards. Then again, it may never really matter.

Reply:
If I take my reasoning one step further, should the rx thread which waits on the rx semaphore, processes the data, yields, and then loops be a standard part of the modem context API?